code/modules/power/stirling.dm

```diff
@@ -0,0 +1,218 @@
+
+	var/power_generated = 0.75*work_done
+	// Excessive power is transferred as heat to the opposite side, reducing efficiency.
+	if(power_generated > max_power)
+		if(reversed)
+			air1_dq += 0.75*(max_power - power_generated)
+		else
+			air2_dq += 0.75*(max_power - power_generated)
+		power_generated = max_power
+		if(prob(5))
+			spark_at(src, cardinal_only = TRUE)
+
+	air1.add_thermal_energy(air1_dq)
+	air2.add_thermal_energy(air2_dq)
+	
+	generate_power(power_generated)
+	lastgen = power_generated
+	genlev = max(0, min(round(4*power_generated / max_power), 4))
+
+	update_icon()
+	update_sound()
+	update_networks()
+
+/obj/machinery/atmospherics/binary/stirling/examine(mob/user, distance)
+	. = ..()
+	if(distance > 1)
+		return
+
+	to_chat(user, "\The [src] is generating [round(lastgen/1000, 0.1)] kW")
+	if(!inserted_cylinder)
+		to_chat(user, "There is no piston cylinder inserted into \the [src].")
+
+/obj/machinery/atmospherics/binary/stirling/attackby(var/obj/item/W, var/mob/user)
+	if((istype(W, /obj/item/tank/stirling)))
+		if(inserted_cylinder)
+			return
+		if(!user.unEquip(W, src))
+			return
+		to_chat(user, SPAN_NOTICE("You insert \the [W] into \the [src]."))
+		inserted_cylinder = W
+		update_icon()
+		return TRUE
+	
+	if(isCrowbar(W) && inserted_cylinder)
+		inserted_cylinder.dropInto(get_turf(src))
+		to_chat(user, SPAN_NOTICE("You remove \the [inserted_cylinder] from \the [src]."))
+		inserted_cylinder = null
+		stop_engine()
+		return TRUE
+
+	if(panel_open && isWrench(W))
+		var/target_frequency = input(user, "Enter the cycle frequency you would like \the [src] to operate at ([MAX_FREQUENCY/4] - [MAX_FREQUENCY] Hz)", "Stirling Frequency", cycle_frequency) as num | null
+		if(!CanPhysicallyInteract(user) || !target_frequency)
+			return
+		cycle_frequency = round(clamp(target_frequency, MAX_FREQUENCY/4, MAX_FREQUENCY))
+		to_chat(usr, SPAN_NOTICE("You adjust \the [src] to operate at a frequency of [cycle_frequency] Hz."))
+		return TRUE
+
+	. = ..()
+
+/obj/machinery/atmospherics/binary/stirling/attack_hand(mob/user)
+	if(!(stat & BROKEN) && use_power != POWER_USE_ACTIVE)
+		if(!inserted_cylinder)
+			to_chat(user, SPAN_WARNING("You must insert a stirling piston cylinder into \the [src] before you can start it!"))
+			return
+		to_chat(user, "You start trying to manually rev up \the [src].")
+		if(do_after(user, 2 SECONDS, src) && use_power != POWER_USE_ACTIVE && inserted_cylinder && !(stat & BROKEN))
+			visible_message("[user] pulls on the starting cord of \the [src], revving it up!", "You pull on the starting cord of \the [src], revving it up!")
+			playsound(src.loc, 'sound/machines/engine.ogg', 35, 1)
+			update_use_power(POWER_USE_ACTIVE)
+		return
+	. = ..()
+
+/obj/machinery/atmospherics/binary/stirling/on_update_icon()
+	cut_overlays()
+	if (stat & (BROKEN))
+		return
+	if (genlev != 0)
+		add_overlay(emissive_overlay('icons/obj/power.dmi', "stirling-op[genlev]")) 
```

Comment:
As you're doing this every tick, you probably do want to track state and only do these operations if the state has changed. There are a bunch of machines that do this (see e.g. APC), but here it's pretty simple, just cache genlev and if genlev == last_genlev do nothing (with e.g. null or -1 corresponding to BROKEN).

Reply:
IMO it's better to clear and regen on call so the function will work if proccalled or from outside contexts where appearance has been manipulated without `genlev` being updated necessarily. Gating the icon update each tick would likely be a better approach.

Reply:
The second solution is what I went with, sound and icons are only updated during process if the gen level has changed.